### PR TITLE
Promise lifecycle events subscriptions and extended debugging/monitoring capabilities (Monitoring for pending promises)

### PR DESCRIPTION
**Pull request by lextiz**
Promises are registred at creation and unregistred when their fate is sealed. At any moment library user can see all pending promises by calling Promise.getPendingPromises() or just the last promises in all chains by calling Promise.getLeafPendingPromises().

### CODE REVIEW

**Comment by lextiz** [comment]
Moved from https://github.com/petkaantonov/bluebird/pull/907

File: src/promise.js [hunk 1]
```diff
@@ -236,6 +236,7 @@ Promise.prototype._then = function (
                 receiver = target === this ? undefined : this._boundTo;
             }
         }
+        this._promiseChained(target);
```
Comment (petkaantonov) [inline src/promise.js]: This should be the original, not `target`. `Target` is when a promise is following another promise:

```js
var b = new Promise(function() {});

var a = new Promise(function(resolve) {
    resolve(b); 
});

//a's target is b

// I chain from a, not b.
a.then(function() {
    
})

```

Comment (lextiz) [inline src/promise.js]: Right, I've fixed this already in promiseChainLengthLimit branch but forgot to fix in this one, thanks. Will fix now.

File: test/mocha/monitor.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		var lastPromise = deferred.promise.thenReturn();
+		assertAllAndLeafPendingPromises([]);
+		return lastPromise;
+	});
+
+	it("promises added to monitor array after creation and removed after direct rejection", function() {
+		var deferred = deferAndMarkAsTestPromise();
+		assertAllAndLeafPendingPromises([deferred.promise]);
+		deferred.resolve();
+		var lastPromise = deferred.promise.thenThrow(new Error("reason")).catch(function (){});
```
Comment (petkaantonov) [inline test/mocha/monitor.js]: Please use caught instead of catch, otherwise test will fail in IE <9

File: test/mocha/monitor.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		G.test = true;
+
+		assertPendingPromises([A, B, C, D, E, F, G], "getPendingPromises");
+		assertPendingPromises([C, E, G], "getLeafPendingPromises");
+
+		var lastPromise = Promise.all([C, E]).then(function() {
+			assertPendingPromises([F, G], "getPendingPromises");
+			assertPendingPromises([G], "getLeafPendingPromises");
+            Fdeferred.reject();
+			return G.catch(function() {
```
Comment (petkaantonov) [inline test/mocha/monitor.js]: catch - > caught

File: test/mocha/monitor.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			assertAllAndLeafPendingPromises([]);
+		});
+		deferred.resolve();
+		return lastPromise;
+	});
+
+	it("promises added to monitor array after creation and removed after rejection", function() {
+		var deferred = deferAndMarkAsTestPromise();
+		assertAllAndLeafPendingPromises([deferred.promise]);
+		var lastPromise = deferred.promise.catch(function() {
```
Comment (petkaantonov) [inline test/mocha/monitor.js]: catch -> caught

File: src/debuggability.js [hunk 5]
```diff
@@ -826,15 +898,20 @@ if (typeof console !== "undefined" && typeof console.warn !== "undefined") {
...
 };
 
 if (longStackTraces) Promise.longStackTraces();
+if (monitor) enableMonitoring();
 
 return {
     longStackTraces: function() {
         return config.longStackTraces;
     },
+    enableMonitoring: function() {
```
Comment (petkaantonov) [inline src/debuggability.js]: is this called from aynwhere?

File: test/mocha/monitor.js (hunk 2 above)
Comment (lextiz) [inline test/mocha/monitor.js]: done

File: test/mocha/monitor.js (hunk 3 above)
Comment (lextiz) [inline test/mocha/monitor.js]: done

File: test/mocha/monitor.js (hunk 4 above)
Comment (lextiz) [inline test/mocha/monitor.js]: done

File: src/debuggability.js (hunk 5 above)
Comment (lextiz) [inline src/debuggability.js]: done

**Comment by lextiz** [comment]
Done with fixes

**Comment by lextiz** [comment]
Added promise life cycle event subscriptions.

**Comment by lextiz** [comment]
Ready for review

**Comment by lextiz** [comment]
@petkaantonov Shall I split the PR to "Promise lifecycle events subscriptions" and "Extended debugging/monitoring capabilities"?

**Comment by petkaantonov** [comment]
The events should be fired on `window`, like in rsvp, so that already implemented tooling works with it

**Comment by lextiz** [comment]
I will add that.

**Comment by lextiz** [comment]
Should I do it in different PR or leave it with debugging stuff?

File: src/extended_debuggability.js [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    }
+    if (!this._tracks) this._tracks = [];
+    if (!next._trackedBy) next._trackedBy = [];
+    this._tracks.push(next);
+    next._trackedBy.push(this);
+}
+
+// Recursive depth of this function is limited so it would produce
+// human-readable input and would not care about circles
+function getDOTGraph (depth, destination) {
```
Comment (petkaantonov) [inline src/extended_debuggability.js]: This stuff is to be implemented by users, the monitor api should only expose low level events and nothing else.

Comment (lextiz) [inline src/extended_debuggability.js]: Ok, at least I had a code review on this part... I have opened new PR for "on" and "off" APIs https://github.com/petkaantonov/bluebird/pull/929.

**Comment by lextiz** [comment]
:pensive:

**Comment by lextiz** [comment]
@petkaantonov I do not want to continue the discussion about this extra debug functionality, but I truly do not understand what are the downsides of having extra debugging functionality as optional part of library. I would be grateful if you could elaborate on reason you do not want to have it.

**Pull request closed without merging**